[개선] 1차 합/불 자동선정 기간 동적으로 변경

leegunwoooo opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

close #209 

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)

> 1차 합/불 자동선정을 동적으로 변경가능하게 했습니다.

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)

- 1차 합/불 선정기간을 동적으로 입력하여 변경할 수 있게 하였습니다.


<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x] 테스트를 완료했나요?
- [x] API 문서를 작성했나요?
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

leegunwoooo committed:
feat(#209) : 1차 불/합 기간을 입력받는 Controller 와 해당하는 Request, Response DTO들 작성

leegunwoooo committed:
refactor(#209) : Controller가 직접 FormScheduler에 의존하던 문제 해결

leegunwoooo committed:
test(#209) : SchedulerController에 대한 테스트 코드 작성 및 MonkBean에 usecase 등록

leegunwoooo committed:
feat(#209) : 입력받는 시간을 전달하는 usecase 작성

leegunwoooo committed:
feat(#209) : 입력받는 시간을 전달하는 usecase 작성

leegunwoooo committed:
refactor(#209) : 컨트롤러 엔드포인트 카멜케이스로 통일

leegunwoooo committed:
refactor(#209) : request dto에 파라미터 없는 기본생성자 추가

leegunwoooo committed:
refactor(#209) : json응답의 형식을 지정

leegunwoooo committed:
test(#209) : 테스트코드의 불필요한 부분 제거

leegunwoooo committed:
Merge branch 'develop' of https://github.com/Bamdoliro/marubase into feat/#209

jyj1289 commented on src/main/java/com/bamdoliro/maru/infrastructure/scheduler/FormScheduler.java:
```diff
@@ -2,17 +2,34 @@
+
+    public void selectFirstPass(LocalDateTime scheduledTime) {
+        if (scheduledTime.isBefore(LocalDateTime.now())) {
+            throw new IllegalArgumentException("이미 지난 시간입니다");
```
P1) IllegalArgumentException은 바인딩 되어있지 않아 예외처리가 제대로 되지 않습니다. 
MaruException을 상속받은 새로운 예외를 만드는 것이 좋지 않을까요?

leegunwoooo replied:
아하 새로운 Exception 클래스를 만들어서 수정하였습니다. 감사합니다

jyj1289 commented on src/main/java/com/bamdoliro/maru/presentation/scheduler/SchedulerController.java:
```diff
@@ -0,0 +1,32 @@
+    private final ScheduleFirstPassUseCase scheduleFirstPassUseCase;
+
+
+    @PostMapping("/updateSchedule")
```
P1) 해당 엔드포인트는 Restful하지 않다고 생각합니다. 
행위를 엔드포인트가 아닌 HTTP 메서드로 표현하는 것이 어떨까요?
또한 엔드포인트 설계시 카멜케이스가 아닌 -를 이용하여 표시합니다. 
ex) update-schedule

leegunwoooo replied:
감사합니다 참고해서 수정했습니당

gimhanul reviewed (approved):
aws에 서버 세팅 어떻게 되어있는지, 인스턴스 한 개에 띄워져 있는지 확인 필요할 거 같습니다!
가용성을 위해 인스턴스 두 개로 돌고 있으면 두 번 실행될 가능성도 있어서요

cabbage16 commented on src/main/java/com/bamdoliro/maru/presentation/scheduler/SchedulerController.java:
```diff
@@ -0,0 +1,32 @@
+    private final ScheduleFirstPassUseCase scheduleFirstPassUseCase;
+
+
+    @PostMapping("/updateSchedule")
```
P4) 엔드포인트에 여러 단어가 들어갈 때는 `-`을 이용하여  /schedule/first-pass와 같은 식으로 하는게 좋을 것 같습니다.

cabbage16 reviewed (changes requested):
동적 스케줄링은 잘 작동하지만, 이렇게 할 경우 서버를 시작할 때마다 계속 요청을 보내서 시간을 정해야하는 불편함이 있어서 초기값이 필요할 거 같습니다.

또한 인스턴스가 2개 이상인 경우 하나의 서버에만 시간이 설정될 수 있어서 만약 멀티 인스턴스 환경으로 운영할 것이라면 다른 방법을 찾아봐야할 것 같습니다.

leegunwoooo committed:
refactor(#209) : 컨트롤러의 엔드포인트 변경

엔드포인트에 여러단어가 들어가기에 엔드포인트를 변경하였습니다.

leegunwoooo committed:
test(#209) : 테스트코드의 엔드포인트 변경

엔드포인트가 변경됨에 따라 테스트코드도 수정하였습니다.

leegunwoooo committed:
feat(#209) : 스케줄을 관리하기 위한 엔티티

스케줄을 db에 저장해 관리하기 위해 엔티티 클래스를 작성하였습니다.

leegunwoooo committed:
feat(#209) : 레포지토리 추가

엔티티 추가함에 따라 레포지토리를 작성하였습니다.

leegunwoooo committed:
feat(#209) : 시간 저장, 업데이트, 불러오기

FormScheduler 에서 입력 받은 시간을 데이터베이스에 업데이트 하도록 수정하였습니다.

db에 저장되어 있는 시간을 불러오는 메서드 작성

leegunwoooo committed:
feat(#209) : 시간 불러오기

서버가 실행될 때 db에 있는 시간을 불러오도록 하였습니다

leegunwoooo committed:
refactor(#209) : ScheduleResponse 반환

UseCase에서 ScheduleResponse를 반환하도록 하였습니다.

leegunwoooo committed:
refactor(#209) : return 문 수정

UseCase에서 Response를 반환해주기 때문에 return 문을 수정하였습니다.

leegunwoooo committed:
test(#209) : 테스트 코드 수정

테스트 코드를 바뀐 코드에 맞게 수정하였습니다.

leegunwoooo committed:
refactor(#209) : 시간 저장

서버가 시작될 때 db에 시간이 저장되도록 하였습니다.

leegunwoooo committed:
refactor(#209) : orElseGet 수정

orElseGet을 통해 얻은 값을 사용하지 않아 통과가 안돼 if문으로 수정

jyj1289 commented on src/main/java/com/bamdoliro/maru/infrastructure/scheduler/FormScheduler.java:
```diff
@@ -27,9 +30,21 @@ public void selectFirstPass(LocalDateTime scheduledTime) {
             scheduledTask.cancel(false);
         }
 
+        FirstPassSchedule schedule = firstPassScheduleRepository.findById(1L)
+                .orElseGet(() -> new FirstPassSchedule(scheduledTime));
+        schedule.update(scheduledTime);
+        firstPassScheduleRepository.save(schedule);
+
```
P3) 튜플 하나를 계속 수정하는 것보다 변경될때마다 DB에 추가하는건 어떨까요? 그렇게하면 로깅도 할 수 있어서 좋을 것 같습니다 :)

leegunwoooo committed:
refactor(#209) : 엔드포인트 변경

- RequestMapping 내의 엔드포인트를 schedules로 변경하였습니다.
- PostMapping의 엔드포인트를 지워 Restful 적으로 변경하였습니다.

leegunwoooo committed:
refactor(#209) : 오타 수정

- 테이블 이름이 schudule로 오타나있던 것을 수정하였습니다.

leegunwoooo committed:
feat(#209) : Exception 추가

- MaruException을 상속받은 Exception 클래스를 작성하였습니다

leegunwoooo committed:
feat(#209) : 스케쥴 Property 추가

- FirstPassScheduleProperty 에 PAST_SCHEDULE 을 추가하여 예외를 발생시키도록 하였습니다.

leegunwoooo committed:
refactor(#209) : Exception 변경

- FormScheduler 의 IllegalArgumentException -> CannotSchedulePastTimeException으로 변경하였습니다

leegunwoooo committed:
refactor(#209) : FirstPassSchedule 엔티티 id 자동 생성 및 update 제거

- 단일 일정에서 업데이트 하는 것이 아닌 새로운 일정을 만들기 때문에 id를 자동생성으로 바꾸었습니다.
- 업데이트 메서드가 필요없어져 update 메서드를 제거하였습니다.

leegunwoooo committed:
refactor(#209) : 파일 구조 변경

- FirstPassScheduleProperty 가 error 디렉토리 안으로 들어가면서 import 문을 변경하였습니다

leegunwoooo committed:
refactor(#209) : 파일 구조 변경

- FirstPassScheduleProperty 가 error 디렉토리 안으로 들어가면서 import 문을 변경하였습니다

leegunwoooo committed:
refactor(#209) : Repository 파일 구조 변경

- FirstPassSchedule가 domain 디렉토리 안으로 들어가 import 문이 변경되었습니다.

leegunwoooo committed:
refactor(#209) : FormScheduler에서 스케쥴 등록시 새로운 객체 생성

- 기존 단일 객체에서 update 를 통해 관리하던 것 insert 되는 방식으로 바뀌어 수정하였습니다

leegunwoooo committed:
feat(#209) : 가장 최근 일정 찾는 메서드 추가

가장 최근에 생성된 일정을 조회하는 JPA 메서드가 필요해 작성하였습니다.

leegunwoooo committed:
refactor(#209) : restoreSchedule를 최신 스케줄 기준으로 변경

오로지 아이디 1번으로만 조회하던 메서드를 최근 생성된 일정을 조회하도록 변경하였습니다.

leegunwoooo committed:
refactor(#209) : ScheduleInitializer 초기값 입력 조건 수정

- 최신 일정을 조회해 없을시 새로 일정을 생성하도록 변경하였습닌다.

leegunwoooo committed:
test(#209) : 엔드포인트 변경

SchedulerController의 엔드포인트가 변경되어 테스트 코드를 수정하였습니다.
